certora/specs/ExchangeRate.spec

```diff
@@ -0,0 +1,92 @@
+
+    assert assets() * sharesBefore < assetsBefore * (shares() + 1);
+}
+
+// Check that if withdraw removed one less share to the user than it does, then the share price would decrease following a withdraw.
+rule boundSharePriceIncreaseOnWithdraw(env e, uint256 assets, address receiver, address onBehalf){
+    require (e.block.timestamp == currentContract.lastUpdate, "assume no interest is accrued");
+
+    mathint assetsBefore = assets();
+    mathint sharesBefore = shares();
+
+    withdraw(e, assets, receiver, onBehalf);
+
+    assert assets() * sharesBefore < assetsBefore * (shares() + 1);
+}
+
+// Check that if mint asks one less asset to the user than it does, then the share price would decrease following a mint.
+rule boundSharePriceIncreaseOnMint(env e, uint256 shares, address onBehalf){
+    require (e.block.timestamp == currentContract.lastUpdate, "assume no interest is accrued");
+
+    mathint assetsBefore = assets();
+    mathint sharesBefore = shares();
+
+    mint(e, shares, onBehalf);
+
+    assert (assets() - 1) * sharesBefore < assetsBefore * shares();
+}
+
+// Check that if redeem gave one more asset to the user than it does, then the share price would decrease following a redeem.
+rule boundSharePriceIncreaseOnRedeem(env e, uint256 shares, address receiver, address onBehalf){
+    require (e.block.timestamp == currentContract.lastUpdate, "assume no interest is accrued");
+
+    mathint assetsBefore = assets();
+    mathint sharesBefore = shares();
+
+    redeem(e, shares, receiver, onBehalf);
+
+    assert (assets() - 1) * sharesBefore < assetsBefore * shares();
+}
+
+// Check that loss realization decreases the share price.
+rule lossRealizationMonotonic(env e, address adapter, bytes data){
+    require (e.block.timestamp == currentContract.lastUpdate, "assume no interest is accrued");
+
+    mathint assetsBefore = assets();
+    mathint sharesBefore = shares();
+
+    accrueInterest(e);
+
+    require (currentContract.lossRealization, "assume loss realization");
+
+    assert assets() * sharesBefore <= assetsBefore * shares();
+}
+
+// Check that share price is increasing, except due to management fees or loss realization.
+rule sharePriceIncreasing(method f, env e, calldataarg a) {
+    require (e.block.timestamp >= currentContract.lastUpdate, "safe requirement because `lastUpdate` is growing and monotonic");
+    require (currentContract.managementFee == 0, "assume management fee to be null");
+    requireInvariant performanceFee();
+
+    require (currentContract.totalSupply > 0, "assume that the vault is seeded");
+
+    requireInvariant balanceOfZero();
+    requireInvariant totalSupplyIsSumOfBalances();
+    requireInvariant virtualSharesBounds();
+
+    mathint assetsBefore = assets();
+    mathint sharesBefore = shares();
+
+    f(e, a);
+
+    require (!currentContract.lossRealization, "assume no loss realization");
+
+    assert assetsBefore * shares() <= assets() * sharesBefore;
```

Comment:
here we would be missing a bug where deposit would make the share price decrease when there has been a loss realization in the accrueInterest. Couldn't we bound the decrease when there is a loss realization (we could record the value instead of a bool).

Reply:
alternatively, we could accrue interest first and then call the functions (probably simpler)

Reply:
Good point. About your ideas:
1. use an integer for the loss realization, I'll try that but maybe that will be too heavy for the prover
2. calling accrueInterest first is a mess because of the firstTotalAssets mecanism. Notably calling it a second time is not easy to see that it's a no-op, in case newTotalAssets was 0 the first time notably

Reply:
In the end 2 was not so bad, because we can use `firstTotalAssets != 0` to make so accruing interest the second time is more clearly a no-op